src/tosyntax/FStar.ToSyntax.Resugar.fs

```diff
@@ -0,0 +1,614 @@
+            let (bv, _) = x.Head in 
+            mk_pat (A.PatVar (bv_as_unique_ident bv, None)), term
+        in
+        if is_pat_app then
+          let args = binders |> List.map (fun (bv, _) -> mk_pat(A.PatVar (bv_as_unique_ident bv, None))) in
+          (mk_pat (A.PatApp (pat, args)), resugar_term term)
+        else
+          (pat, resugar_term term)
+      in
+      let bnds = List.map (resugar_one_binding) bnds in
+      let body = resugar_term body in
+      mk (A.Let((if is_rec then A.Rec else A.NoLetQualifier), bnds, body))
+           
+    | Tm_uvar _ ->
+      failwith "This case is impossible since it is not created in desugar"
+
+    | Tm_meta(e, m) ->
+       let resugar_meta_desugared = function
+          | Data_app ->
+              begin match (SS.compress e).n with
+                | Tm_app(head, args) ->
+                    let rec aux h = match ((SS.compress h).n) with
+                      | Tm_fvar fv -> lid_of_fv fv, []
+                      | Tm_uinst(h, u) -> 
+                        let h, l = aux h in
+                        h, l@u
+                      | _ -> failwith "wrong Data_app head format" 
+                    in
+                    let head, universes = aux head in
+                    let universes = List.map (fun u -> (resugar_universe u t.pos, A.UnivApp)) universes in
+                    let args = List.map (fun (t, _) -> (resugar_term t, A.Nothing)) args in
+                    mk (A.Construct(head, args@universes))
+                | _ -> failwith "wrong Data_app format"
+              end
+          | Sequence ->
+              let term = resugar_term e in
+              resugar_sequence term
+          | Primop (* doesn't seem to be generated by desugar *)
+          | Masked_effect (* doesn't seem to be generated by desugar *)
+          | Meta_smt_pat -> (* nothing special, just resugar the term *)
+              resugar_term e
+          | Mutable_alloc ->
+              let term = resugar_term e in
+              begin match term.tm with
+                | A.Let(A.NoLetQualifier,l,t) -> mk (A.Let(A.Mutable, l, t))
+                | _ -> failwith "mutable_alloc should have let term with no qualifier"
+              end
+          | Mutable_rval ->
+              let fv = S.lid_as_fv C.sread_lid Delta_constant None in
+              begin match (SS.compress e).n with 
+                | Tm_app({n=Tm_fvar fv}, [(term, _)])-> resugar_term term
+                | _ -> failwith "mutable_rval should have app term"
+              end
+      in
+      begin match m with
+      | Meta_pattern pats ->
+          failwith "This case is impossible, if Forall and Exist is properly resugared"
+      | Meta_labeled (l, _, p) ->
+          mk (A.Labeled(resugar_term e, l, p))
+      | Meta_desugared i -> 
+          resugar_meta_desugared i
+      | Meta_named _ 
+      | Meta_monadic _ 
+      | Meta_monadic_lift _ -> 
+          failwith "cases not generated by desugar"
```

Comment:
Removing these cases is not a solution either. If you have a look at `print_term` in `src/syntax/FStar.Syntax.Print.fs` you can see that dependending on some options we are either removing all meta informations (using `unmeta`) or printing them. The way we print these is not valid F* syntax and we need to come up with a way to provide the information into the surface AST. I think that a first attempt would be to use an `Ascribed` node for `Meta_monadic` and `Meta_monadic_lift`. In more details if you have a term `Meta(e, Meta_monadic(t, m))` or `Meta(e, Meta_monadic_lift(t, _, m))` you should construct a term along the lines of `Ascribed( e', Constructor(Name m, [t']))` where e' is the translation of e and t' the translation of t.